File: externals/Async-CTP/samples/VB/Dataflow/Real/Estate/RealtorAgent.vb
'--------------------------------------------------------------------------
' 
'  Copyright (c) Microsoft Corporation.  All rights reserved. 
' 
'  File: RealtorAgent.vb
'
'--------------------------------------------------------------------------

Imports System.Threading.Tasks.Dataflow
Imports System.Threading.Tasks

Friend Class RealtorAgent
    ''' <summary>Gets a source for querying the realtor's current buyers.</summary>
    Public ReadOnly CurrentBuyers As New BroadcastBlock(Of String())(Function(i) i)
    ''' <summary>Gets a source for querying the realtor's current listings.</summary>
    Public ReadOnly CurrentListings As New BroadcastBlock(Of Listing())(Function(i) i)

    ''' <summary>Gets a target for new listings the realtor should represent.</summary>
    Private ReadOnly m_listHouse As ITargetBlock(Of Listing)
    ''' <summary>Gets a target for alerting the realtor to a new buyer.</summary>
    Private ReadOnly m_subscribeBuyer As ITargetBlock(Of BuyerAgent)
    ''' <summary>Gets a target for querying the realtor for listings that match certain criteria..</summary>
    Private ReadOnly m_search As ITargetBlock(Of Tuple(Of SearchFilter, WriteOnceBlock(Of Listing)))
    ''' <summary>Gets a target for making an offer to the realtor to buy a particular listing.</summary>
    Private ReadOnly m_buy As ITargetBlock(Of Tuple(Of Offer, WriteOnceBlock(Of Boolean)))

    ''' <summary>A collection of the realtor's listings.</summary>
    Private ReadOnly m_listings As New HashSet(Of Listing)
    ''' <summary>A collection of the realtor's buyers.</summary>
    Private ReadOnly m_buyers As New HashSet(Of BuyerAgent)

    ''' <summary>Initializes the realtor agent.</summary>
    Public Sub New()
        ' Allow certain realtor processing to happen concurrently and some processing to happen exclusively
        Dim ce = New ConcurrentExclusiveSchedulerPair()
        Dim exclusive = New ExecutionDataflowBlockOptions With {.TaskScheduler = ce.ExclusiveScheduler}
        Dim concurrent = New ExecutionDataflowBlockOptions With {.TaskScheduler = ce.ConcurrentScheduler}

        ' When a new buyer arrives, notify them of any properties we may have
        m_subscribeBuyer = New ActionBlock(Of BuyerAgent)(
            Sub(buyer)
                m_buyers.Add(buyer)
                CurrentBuyers.Post(m_buyers.Select(Function(b) b.Name).ToArray())
                If (m_listings.Count > 0) Then buyer.AvailableRealtor.Post(Me)
            End Sub, exclusive)

        ' When a buyer queries us about potential listings that match their criteria,
        ' search and let them know if we found any
        m_search = New ActionBlock(Of Tuple(Of SearchFilter, WriteOnceBlock(Of Listing)))(
            Sub(requestResponse)
                Dim s = requestResponse.Item1
                For Each listing In m_listings
                    If (listing.Price >= s.MinPrice AndAlso listing.Price <= s.MaxPrice) Then
                        requestResponse.Item2.Post(listing)
                        Return
                    End If
                Next
                requestResponse.Item2.Post(Nothing)
            End Sub, concurrent)

        ' When a new listing is presented to us, notify all buyers that a new
        ' listing is available.
        m_listHouse = New ActionBlock(Of Listing)(
            Sub(listing)
                m_listings.Add(listing)
                CurrentListings.Post(m_listings.ToArray())
                If (m_buyers.Count > 0) Then
                    For Each buyer In m_buyers
                        buyer.AvailableRealtor.Post(Me)
                    Next
                End If
            End Sub, exclusive)

        ' When an offer is made, if it meets the listing and the listing is still
        ' available, sell.  Otherwise, reject.
        m_buy = New ActionBlock(Of Tuple(Of Offer, WriteOnceBlock(Of Boolean)))(
            Sub(requestResponse)
                Dim offer = requestResponse.Item1
                If (m_listings.Contains(offer.Listing) AndAlso
                    offer.Listing.Price <= offer.Price AndAlso
                    m_buyers.Contains(offer.Buyer)) Then
                    m_buyers.Remove(offer.Buyer)
                    m_listings.Remove(offer.Listing)

                    requestResponse.Item2.Post(True)

                    CurrentBuyers.Post(m_buyers.Select(Function(b) b.Name).ToArray())
                    CurrentListings.Post(m_listings.ToArray())
                Else
                    requestResponse.Item2.Post(False)
                End If
            End Sub, exclusive)
    End Sub

    ''' <summary>Asynchronously searches based on the specified filter criteria.</summary>
    ''' <param name="filter">The criteria with which to search.</param>
    ''' <returns>A task representing the result of the search.</returns>
    Public Function SearchAsync(filter As SearchFilter) As Task(Of Listing)
        Dim response = New WriteOnceBlock(Of Listing)(Function(i) i)
        m_search.Post(Tuple.Create(filter, response))
        Return response.ReceiveAsync()
    End Function

    ''' <summary>Asynchronously makes an offer.</summary>
    ''' <param name="offer">The offer to make</param>
    ''' <returns>A task representing whether the offer was accepted.</returns>
    Public Function MakeOfferAsync(offer As Offer) As Task(Of Boolean)
        Dim response = New WriteOnceBlock(Of Boolean)(Function(i) i)
        m_buy.Post(Tuple.Create(offer, response))
        Return response.ReceiveAsync()
    End Function

    ''' <summary>List a new house.</summary>
    ''' <param name="listing">The listing.</param>
    Public Sub ListHouse(listing As Listing)
        m_listHouse.Post(listing)
    End Sub

    ''' <summary>Take on a new client.</summary>
    ''' <param name="buyer">The buyer client.</param>
    Public Sub ServeBuyer(buyer As BuyerAgent)
        m_subscribeBuyer.Post(buyer)
    End Sub
End Class

File: externals/Async-CTP/samples/CS/Dataflow/Real/Estate/RealtorAgent.cs
//--------------------------------------------------------------------------
// 
//  Copyright (c) Microsoft Corporation.  All rights reserved. 
// 
//  File: RealtorAgent.cs
//
//--------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Threading.Tasks.Dataflow;

namespace RealEstateSimulation
{
    /// <summary>Represents a realtor in the real estate simulation.</summary>
    internal class RealtorAgent
    {
        /// <summary>Gets a source for querying the realtor's current buyers.</summary>
        public BroadcastBlock<string[]> CurrentBuyers { get; private set; }
        /// <summary>Gets a source for querying the realtor's current listings.</summary>
        public BroadcastBlock<Listing[]> CurrentListings { get; private set; }

        /// <summary>Gets a target for new listings the realtor should represent.</summary>
        private readonly ITargetBlock<Listing> m_listHouse;
        /// <summary>Gets a target for alerting the realtor to a new buyer.</summary>
        private readonly ITargetBlock<BuyerAgent> m_subscribeBuyer;
        /// <summary>Gets a target for querying the realtor for listings that match certain criteria..</summary>
        private readonly ITargetBlock<Tuple<SearchFilter, WriteOnceBlock<Listing>>> m_search;
        /// <summary>Gets a target for making an offer to the realtor to buy a particular listing.</summary>
        private readonly ITargetBlock<Tuple<Offer, WriteOnceBlock<bool>>> m_buy;

        /// <summary>A collection of the realtor's listings.</summary>
        private readonly HashSet<Listing> m_listings = new HashSet<Listing>();
        /// <summary>A collection of the realtor's buyers.</summary>
        private readonly HashSet<BuyerAgent> m_buyers = new HashSet<BuyerAgent>();

        /// <summary>Initializes the realtor agent.</summary>
        public RealtorAgent()
        {
            // Allow certain realtor processing to happen concurrently and some processing to happen exclusively
            var ce = new ConcurrentExclusiveSchedulerPair();
            var exclusive = new ExecutionDataflowBlockOptions { TaskScheduler = ce.ExclusiveScheduler };
            var concurrent = new ExecutionDataflowBlockOptions { TaskScheduler = ce.ConcurrentScheduler };

            // When a new buyer arrives, notify them of any properties we may have
            m_subscribeBuyer = new ActionBlock<BuyerAgent>(buyer =>
            {
                m_buyers.Add(buyer);
                CurrentBuyers.Post(m_buyers.Select(b => b.Name).ToArray());
                if (m_listings.Count > 0) buyer.AvailableRealtor.Post(this);
            }, exclusive);

            // When a buyer queries us about potential listings that match their criteria,
            // search and let them know if we found any
            m_search = new ActionBlock<Tuple<SearchFilter, WriteOnceBlock<Listing>>>(requestResponse =>
            {
                var s = requestResponse.Item1;
                foreach (var listing in m_listings)
                {
                    if (listing.Price >= s.MinPrice && listing.Price <= s.MaxPrice)
                    {
                        requestResponse.Item2.Post(listing);
                        return;
                    }
                }
                requestResponse.Item2.Post(null);
            }, concurrent);

            // When a new listing is presented to us, notify all buyers that a new
            // listing is available.
            m_listHouse = new ActionBlock<Listing>(listing =>
            {
                m_listings.Add(listing);
                CurrentListings.Post(m_listings.ToArray());
                if (m_buyers.Count > 0)
                {
                    foreach (var buyer in m_buyers) buyer.AvailableRealtor.Post(this);
                }
            }, exclusive);

            // When an offer is made, if it meets the listing and the listing is still
            // available, sell.  Otherwise, reject.
            m_buy = new ActionBlock<Tuple<Offer, WriteOnceBlock<bool>>>(requestResponse =>
            {
                var offer = requestResponse.Item1;
                if (m_listings.Contains(offer.Listing) &&
                    offer.Listing.Price <= offer.Price &&
                    m_buyers.Contains(offer.Buyer))
                {
                    m_buyers.Remove(offer.Buyer);
                    m_listings.Remove(offer.Listing);

                    requestResponse.Item2.Post(true);

                    CurrentBuyers.Post(m_buyers.Select(b => b.Name).ToArray());
                    CurrentListings.Post(m_listings.ToArray());
                }
                else requestResponse.Item2.Post(false);
            }, exclusive);

            CurrentBuyers = new BroadcastBlock<string[]>(_ => _);
            CurrentListings = new BroadcastBlock<Listing[]>(_ => _);
        }

        /// <summary>Asynchronously searches based on the specified filter criteria.</summary>
        /// <param name="filter">The criteria with which to search.</param>
        /// <returns>A task representing the result of the search.</returns>
        public Task<Listing> SearchAsync(SearchFilter filter)
        {
            var response = new WriteOnceBlock<Listing>(_ => _);
            m_search.Post(Tuple.Create(filter, response));
            return response.ReceiveAsync();
        }

        /// <summary>Asynchronously makes an offer.</summary>
        /// <param name="offer">The offer to make</param>
        /// <returns>A task representing whether the offer was accepted.</returns>
        public Task<bool> MakeOfferAsync(Offer offer)
        {
            var response = new WriteOnceBlock<bool>(_ => _);
            m_buy.Post(Tuple.Create(offer, response));
            return response.ReceiveAsync();
        }

        /// <summary>List a new house.</summary>
        /// <param name="listing">The listing.</param>
        public void ListHouse(Listing listing) { m_listHouse.Post(listing); }

        /// <summary>Take on a new client.</summary>
        /// <param name="buyer">The buyer client.</param>
        public void ServeBuyer(BuyerAgent buyer) { m_subscribeBuyer.Post(buyer); }
    }
}